File: mobile/src/main/scala/com/github/shadowsocks/acl/Acl.scala
package com.github.shadowsocks.acl

import java.io.{File, FileNotFoundException}
import java.net.URL

import android.util.Log
import com.github.shadowsocks.ShadowsocksApplication.app
import com.github.shadowsocks.utils.IOUtils
import com.j256.ormlite.field.DatabaseField

import scala.collection.mutable
import scala.io.Source

/**
  * ACL handler compliant with: src/main/jni/shadowsocks-libev/src/acl.c
  *
  * OrmLite integration is unused for now.
  *
  * @author Mygod
  */
class Acl {
  import Acl._

  @DatabaseField(generatedId = true)
  var id: Int = _
  val bypassHostnames = new mutable.SortedList[String]()
  val proxyHostnames = new mutable.SortedList[String]()
  val subnets = new mutable.SortedList[Subnet]()
  val urls = new mutable.SortedList[URL]()(urlOrdering)

  @DatabaseField
  var bypass: Boolean = _

  def getBypassHostnamesString: String = bypassHostnames.mkString("\n")
  def getProxyHostnamesString: String = proxyHostnames.mkString("\n")
  def getSubnetsString: String = subnets.mkString("\n")
  def getUrlsString: String = urls.mkString("\n")
  def setBypassHostnamesString(value: String) {
    bypassHostnames.clear()
    bypassHostnames ++= value.split("\n")
  }
  def setProxyHostnamesString(value: String) {
    proxyHostnames.clear()
    proxyHostnames ++= value.split("\n")
  }
  def setSubnetsString(value: String) {
    subnets.clear()
    subnets ++= value.split("\n").map(Subnet.fromString)
  }
  def setUrlsString(value: String) {
    urls.clear()
    urls ++= value.split("\n").map(new URL(_))
  }

  def fromAcl(other: Acl): Acl = {
    bypassHostnames.clear()
    bypassHostnames ++= other.bypassHostnames
    proxyHostnames.clear()
    proxyHostnames ++= other.proxyHostnames
    subnets.clear()
    subnets ++= other.subnets
    urls.clear()
    urls ++= other.urls
    bypass = other.bypass
    this
  }
  def fromSource(value: Source, defaultBypass: Boolean = false): Acl = {
    bypassHostnames.clear()
    proxyHostnames.clear()
    this.subnets.clear()
    urls.clear()
    bypass = defaultBypass
    lazy val bypassSubnets = new mutable.SortedList[Subnet]()
    lazy val proxySubnets = new mutable.SortedList[Subnet]()
    var hostnames: mutable.SortedList[String] = if (defaultBypass) proxyHostnames else bypassHostnames
    var subnets: mutable.SortedList[Subnet] = if (defaultBypass) proxySubnets else bypassSubnets
    for (line <- value.getLines()) (line.indexOf('#') match {
      case -1 => line
      case index =>
        line.substring(index + 1) match {
          case networkAclParser(url) => urls.add(new URL(url))
          case _ => // ignore
        }
        line.substring(0, index)  // trim comments
    }).trim match {
      case "[outbound_block_list]" =>
        hostnames = null
        subnets = null
      case "[black_list]" | "[bypass_list]" =>
        hostnames = bypassHostnames
        subnets = bypassSubnets
      case "[white_list]" | "[proxy_list]" =>
        hostnames = proxyHostnames
        subnets = proxySubnets
      case "[reject_all]" | "[bypass_all]" => bypass = true
      case "[accept_all]" | "[proxy_all]" => bypass = false
      case input if subnets != null && input.nonEmpty => try subnets += Subnet.fromString(input) catch {
        case _: IllegalArgumentException => hostnames += input
      }
      case _ =>
    }
    this.subnets ++= (if (bypass) proxySubnets else bypassSubnets)
    this
  }
  final def fromId(id: String): Acl = fromSource(Source.fromFile(Acl.getFile(id)))

  def flatten(depth: Int): Acl = {
    if (depth > 0) for (url <- urls) {
      val child = new Acl().fromSource(Source.fromURL(url), bypass).flatten(depth - 1)
      if (bypass != child.bypass) {
        Log.w(TAG, "Imported network ACL has a conflicting mode set. This will probably not work as intended. URL: %s"
          .format(url))
        child.subnets.clear() // subnets for the different mode are discarded
        child.bypass = bypass
      }
      bypassHostnames ++= child.bypassHostnames
      proxyHostnames ++= child.proxyHostnames
      subnets ++= child.subnets
    }
    urls.clear()
    this
  }

  override def toString: String = {
    val result = new StringBuilder()
    result.append(if (bypass) "[bypass_all]\n" else "[proxy_all]\n")
    val (bypassList, proxyList) =
      if (bypass) (bypassHostnames.toStream, subnets.toStream.map(_.toString) #::: proxyHostnames.toStream)
      else (subnets.toStream.map(_.toString) #::: bypassHostnames.toStream, proxyHostnames.toStream)
    if (bypassList.nonEmpty) {
      result.append("[bypass_list]\n")
      result.append(bypassList.mkString("\n"))
      result.append('\n')
    }
    if (proxyList.nonEmpty) {
      result.append("[proxy_list]\n")
      result.append(proxyList.mkString("\n"))
      result.append('\n')
    }
    result.append(urls.map("#IMPORT_URL <%s>\n".format(_)).mkString)
    result.toString
  }

  // Don't change: dummy fields for OrmLite interaction

  // noinspection ScalaUnusedSymbol
  @DatabaseField(useGetSet = true)
  private val bypassHostnamesString: String = null
  // noinspection ScalaUnusedSymbol
  @DatabaseField(useGetSet = true)
  private val proxyHostnamesString: String = null
  // noinspection ScalaUnusedSymbol
  @DatabaseField(useGetSet = true)
  private val subnetsString: String = null
}

object Acl {
  final val TAG = "Acl"
  final val ALL = "all"
  final val BYPASS_LAN = "bypass-lan"
  final val BYPASS_CHN = "bypass-china"
  final val BYPASS_LAN_CHN = "bypass-lan-china"
  final val GFWLIST = "gfwlist"
  final val CHINALIST = "china-list"
  final val CUSTOM_RULES = "custom-rules"
  final val CUSTOM_RULES_FLATTENED = "custom-rules-flattened"
  private val networkAclParser = "^IMPORT_URL\\s*<(.+)>\\s*$".r

  private val urlOrdering: Ordering[URL] = Ordering.by(url => (url.getHost, url.getPort, url.getFile, url.getProtocol))

  def getFile(id: String) = new File(app.getFilesDir, id + ".acl")
  def customRules: Acl = {
    val acl = new Acl()
    try acl.fromId(CUSTOM_RULES) catch {
      case _: FileNotFoundException =>
    }
    acl.bypass = true
    acl.bypassHostnames.clear() // everything is bypassed
    acl
  }
  def save(id: String, acl: Acl): Unit = IOUtils.writeString(getFile(id), acl.toString)
}

File: mobile/src/main/java/com/github/shadowsocks/acl/Acl.kt
/*******************************************************************************
 *                                                                             *
 *  Copyright (C) 2017 by Max Lv <max.c.lv@gmail.com>                          *
 *  Copyright (C) 2017 by Mygod Studio <contact-shadowsocks-android@mygod.be>  *
 *                                                                             *
 *  This program is free software: you can redistribute it and/or modify       *
 *  it under the terms of the GNU General Public License as published by       *
 *  the Free Software Foundation, either version 3 of the License, or          *
 *  (at your option) any later version.                                        *
 *                                                                             *
 *  This program is distributed in the hope that it will be useful,            *
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of             *
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the              *
 *  GNU General Public License for more details.                               *
 *                                                                             *
 *  You should have received a copy of the GNU General Public License          *
 *  along with this program. If not, see <http://www.gnu.org/licenses/>.       *
 *                                                                             *
 *******************************************************************************/

package com.github.shadowsocks.acl

import android.content.Context
import android.support.v7.util.SortedList
import android.util.Log
import com.github.shadowsocks.App.Companion.app
import com.github.shadowsocks.preference.DataStore
import com.github.shadowsocks.utils.Subnet
import com.github.shadowsocks.utils.asIterable
import java.io.File
import java.io.IOException
import java.io.Reader
import java.net.URL

class Acl {
    companion object {
        const val TAG = "Acl"
        const val ALL = "all"
        const val BYPASS_LAN = "bypass-lan"
        const val BYPASS_CHN = "bypass-china"
        const val BYPASS_LAN_CHN = "bypass-lan-china"
        const val GFWLIST = "gfwlist"
        const val CHINALIST = "china-list"
        const val CUSTOM_RULES = "custom-rules"

        val networkAclParser = "^IMPORT_URL\\s*<(.+)>\\s*$".toRegex()

        fun getFile(id: String, context: Context = app.deviceContext) = File(context.filesDir, id + ".acl")

        var customRules: Acl
            get() {
                val acl = Acl()
                val str = DataStore.publicStore.getString(CUSTOM_RULES)
                if (str != null) {
                    acl.fromReader(str.reader())
                    if (!acl.bypass) {
                        acl.subnets.clear()
                        acl.hostnames.clear()
                        acl.bypass = true
                    }
                } else acl.bypass = true
                return acl
            }
            set(value) = DataStore.publicStore.putString(CUSTOM_RULES, if ((!value.bypass ||
                    value.subnets.size() == 0 && value.hostnames.size() == 0) && value.urls.size() == 0)
                null else value.toString())
        fun save(id: String, acl: Acl) = getFile(id).writeText(acl.toString())
    }

    private abstract class BaseSorter<T> : SortedList.Callback<T>() {
        override fun onInserted(position: Int, count: Int) { }
        override fun areContentsTheSame(oldItem: T?, newItem: T?): Boolean = oldItem == newItem
        override fun onMoved(fromPosition: Int, toPosition: Int) { }
        override fun onChanged(position: Int, count: Int) { }
        override fun onRemoved(position: Int, count: Int) { }
        override fun areItemsTheSame(item1: T?, item2: T?): Boolean = item1 == item2
        override fun compare(o1: T?, o2: T?): Int =
                if (o1 == null) if (o2 == null) 0 else 1 else if (o2 == null) -1 else compareNonNull(o1, o2)
        abstract fun compareNonNull(o1: T, o2: T): Int
    }
    private open class DefaultSorter<T : Comparable<T>> : BaseSorter<T>() {
        override fun compareNonNull(o1: T, o2: T): Int = o1.compareTo(o2)
    }
    private object StringSorter : DefaultSorter<String>()
    private object SubnetSorter : DefaultSorter<Subnet>()
    private object URLSorter : BaseSorter<URL>() {
        private val ordering = compareBy<URL>({ it.host }, { it.port }, { it.file }, { it.protocol })
        override fun compareNonNull(o1: URL, o2: URL): Int = ordering.compare(o1, o2)
    }

    val hostnames = SortedList(String::class.java, StringSorter)
    val subnets = SortedList(Subnet::class.java, SubnetSorter)
    val urls = SortedList(URL::class.java, URLSorter)
    var bypass = false

    fun clear(): Acl {
        hostnames.clear()
        subnets.clear()
        urls.clear()
        return this
    }

    fun fromAcl(other: Acl): Acl {
        clear()
        for (item in other.hostnames.asIterable()) hostnames.add(item)
        for (item in other.subnets.asIterable()) subnets.add(item)
        for (item in other.urls.asIterable()) urls.add(item)
        bypass = other.bypass
        return this
    }
    fun fromReader(reader: Reader, defaultBypass: Boolean = false): Acl {
        clear()
        bypass = defaultBypass
        val proxyHostnames by lazy { SortedList(String::class.java, StringSorter) }
        val bypassHostnames by lazy { SortedList(String::class.java, StringSorter) }
        val bypassSubnets by lazy { SortedList(Subnet::class.java, SubnetSorter) }
        val proxySubnets by lazy { SortedList(Subnet::class.java, SubnetSorter) }
        var hostnames: SortedList<String>? = if (defaultBypass) proxyHostnames else bypassHostnames
        var subnets: SortedList<Subnet>? = if (defaultBypass) proxySubnets else bypassSubnets
        reader.useLines {
            for (line in it) {
                @Suppress("PLATFORM_CLASS_MAPPED_TO_KOTLIN")
                val blocks = (line as java.lang.String).split("#", 2)
                val url = networkAclParser.matchEntire(blocks.getOrElse(1, {""}))?.groupValues?.getOrNull(1)
                if (url != null) urls.add(URL(url))
                val input = blocks[0].trim()
                when (input) {
                    "[outbound_block_list]" -> {
                        hostnames = null
                        subnets = null
                    }
                    "[black_list]", "[bypass_list]" -> {
                        hostnames = bypassHostnames
                        subnets = bypassSubnets
                    }
                    "[white_list]", "[proxy_list]" -> {
                        hostnames = proxyHostnames
                        subnets = proxySubnets
                    }
                    "[reject_all]", "[bypass_all]" -> bypass = true
                    "[accept_all]", "[proxy_all]" -> bypass = false
                    else -> if (subnets != null && input.isNotEmpty()) {
                        val subnet = Subnet.fromString(input)
                        if (subnet == null) hostnames!!.add(input) else subnets!!.add(subnet)
                    }
                }
            }
        }
        for (item in (if (bypass) proxyHostnames else bypassHostnames).asIterable()) this.hostnames.add(item)
        for (item in (if (bypass) proxySubnets else bypassSubnets).asIterable()) this.subnets.add(item)
        return this
    }

    fun fromId(id: String): Acl = try {
        fromReader(Acl.getFile(id).bufferedReader())
    } catch (_: IOException) { this }

    fun flatten(depth: Int): Acl {
        if (depth > 0) for (url in urls.asIterable()) {
            val child = Acl()
            try {
                child.fromReader(url.openStream().bufferedReader(), bypass).flatten(depth - 1)
            } catch (_: IOException) {
                continue
            }
            if (bypass != child.bypass) {
                Log.w(TAG, "Imported network ACL has a conflicting mode set. " +
                        "This will probably not work as intended. URL: $url")
                // rules for the different mode are discarded
                child.hostnames.clear()
                child.subnets.clear()
                child.bypass = bypass
            }
            for (item in child.hostnames.asIterable()) hostnames.add(item)
            for (item in child.subnets.asIterable()) subnets.add(item)
        }
        urls.clear()
        return this
    }

    override fun toString(): String {
        val result = StringBuilder()
        result.append(if (bypass) "[bypass_all]\n[proxy_list]\n" else "[proxy_all]\n[bypass_list]\n")
        result.append(subnets.asIterable().joinToString("\n"))
        result.append('\n')
        result.append(hostnames.asIterable().joinToString("\n"))
        result.append('\n')
        result.append(urls.asIterable().joinToString("") { "#IMPORT_URL <$it>\n" })
        return result.toString()
    }
}
